### PR TITLE
Feature/search

### LINKED ISSUE #10: MVP- Search Bar component
As a user I should see a search bar component under the header at all times.  Upon typing and clicking search or pressing enter, I should see a page of results.  Results are a list of cards, a card has a title, artist name, and an cover img.

### PR DESCRIPTION
**Pull request by alivaditis**
This PR will close #10 

## What I did:

Component for Search and Results using routes.

States: results, pagination
useParams: query, page

Search navigates to a search route with dynamic query value.
Results calls the api with the query and page provided from useParams, sets results state and pagination state from the results of the api call.
On 'last' or 'next' button click browser navigates to a /search/:query/:page provides, provides either page-1 or page+1 for page param.

## Testing Notes/Extras: 
(Include steps for happy path user flows for all feature work added in this change)

Tests included for search by either clicking search button or pressing enter.
Tests included for pressing last and next button.
Test included for no results for query.

(Include edge cases/potential sad paths to look out for)

Did this break anything?

- [x]  No
- [ ]  Yes

Type of change

- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  Refactor(DRY-ing up/ reorganizing code, etc.)
- [ ]  Super small fix (Corrected a typo, removed a comment, etc.)
- [ ]  Skip all the other stuff and briefly explain the fix.

Checklist:

- [x]  If this code needs to be tested, all tests are passing
- [x]  The code runs locally
- [ ]  There are comments where clarification/ organization is needed
- [x]  The code is DRY. If it's not, I tried my best
- [x]  I reviewed my code before pushing

What's next?
Specific Album Page, A way of handling errors across components.

### CODE REVIEW

File: src/api.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+function searchAlbums(query, page) {
+  return fetch(`https://api.discogs.com/database/search?q=${query}&type=master&key=mbubAaAXseWPUpaJLkKU&secret=TrELhUezCNdFoIfmoAdHZmfJIXljOSfW&format=vinyl&page=${page}`)
+    .then(res => {
+      if (!res.ok) {
+      throw new Error('Albums not found')
+      }
+      return res.json()})
```
Comment (jalbe0076) [inline src/api.js:7]: We should pull this out and make it into its own function in case we have to have another fetch call. This would make it more dynamic.
You could the call the function in the then block `.then(res => handleError(res)` 
In addition, I think the message could be made more informative. If the server where it returns a 500 error or if the album is not found a 404 error we can let the user know. This can also be discussed. `throw new Error('HTTP Error: ${res.status} -- Please try again later')`

File: src/Components/App/App.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import Results from '../Results/Results'
+import Album from '../Album/Album'
+import './App.css';
+
+function App() {
+  const [query, setQuery] = useState('')
+  
+  return (
+    <div className="App">
+      <Search setQuery={setQuery}/>
```
Comment (jalbe0076) [inline src/Components/App/App.js:13]: I've gotten comments saying this is not best practice, It is generally considered better practice to handle state management within the component that owns the state, rather than passing setter functions from outside the component.

File: src/Components/Results/Results.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  useEffect(() => {
+    setLoading(true)
+    searchAlbums(query, 1)
+      .then(data => {
+        setResults(data.results)
+        setPagination(data.pagination)
+        setLoading(false)
+      })
+      .catch(error => alert(error.message))
```
Comment (jalbe0076) [inline src/Components/Results/Results.js:19]: Displaying this on the DOM instead of having an alert box pop up would be less annoying for the user. We could easily have an error state and have a conditional if true, display the `error.message`

File: src/Components/Results/Results.css [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  color: white;
+  padding-top: 60px;
+}
+
+.results--none {
+  justify-content: center;
+  width: 100%;
+}
+
+.results--grid {
```
Comment (jalbe0076) [inline src/Components/Results/Results.css:20]: this breaks around 300px wide. We should drop down to 1 album per line at smaller sizes or figure out how to deal with the white line on the right and the partially cut text on a 280px screen size.

File: src/Components/Results/Results.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    !isLoading ? (<section className='results'>
+      {results.length ? (
+        <div className='results--grid'>
+            {results.map(result => (
+              <Card key={result.masterId} result={result}/>
+            ))}
+        </div>
+        ) : <div className='results--none'><p>No results for {query}</p></div>}
+      {results.length && <div className='results--pages'>
+        {pagination.page > 1 && <button className='results--last'onClick={() => handleClick('last')}>last</button>}
```
Comment (jalbe0076) [inline src/Components/Results/Results.js:55]: Changing last to previous makes more sense to me. When I click the last button I expect to go to the last page. So if there is 1 of 200, clicking last would take me to 200

File: src/Components/Results/Results.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        setResults(data.results)
+        setPagination(data.pagination)
+        setLoading(false)
+      })
+      .catch(error => alert(error.message))
+  }, [ query ])
+
+  const handleClick = (direction) => {
+    setLoading(true)
+    if (direction === 'last') {
```
Comment (jalbe0076) [inline src/Components/Results/Results.js:24]: We can save the value of `pagination.page - 1` or `pagination.page + 1` to a variable using a conditional to pass as the second argument when the `searchAlbums` function is invoked. This way you only need to call it once and you can make the code more DRY

File: src/Components/Search/search.png
Comment (jalbe0076) [inline src/Components/Search/search.png:1]: images should be grouped together in the public folder, I made an images folder to group them together.

File: src/Components/Search/Search.css [hunk 7]
```diff
@@ -0,0 +1,29 @@
+.search--container {
+  position: relative;
+  display: flex;
+  align-items: center;
+}
+
+.search--input {
+  width: 100%;
```
Comment (jalbe0076) [inline src/Components/Search/Search.css:8]: Is it intentional to have the search bar extend the whole width of the screen? It looks really odd on my large screens. We could discuss this.

**Review by jalbe0076 — CHANGES_REQUESTED** [review]
Great job Alec, looks like the data we got back from the fetch was a little more complicated than I expected.

File: cypress/e2e/search.cy.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      .get('.results--card').eq(1).contains('p', 'Stankonia')
+      .get('.results--card').eq(1).contains('p', 'OutKast')
+  })
+
+  it('As a user, if there are no results I should see a no results for query message', () => {
+    cy.get('.search--form').should('be.visible')
+      .get('.search--input').type('awefdfi')
+      .get('.search--button').click()
+      .get('.results--none').should('be.visible')
+      .get('.results--none').contains('awefdfi')
```
[the remark this answers is not in the document]
Comment (alivaditis) [reply cypress/e2e/search.cy.js:69]: good catch, made a new fixture, noresults.json

File: src/Components/Results/Results.js (hunk 5 above)
Comment (alivaditis) [reply src/Components/Results/Results.js]: ah that makes sense, changed it

File: src/Components/Results/Results.js (hunk 6 above)
Comment (alivaditis) [reply src/Components/Results/Results.js:28]: true, implemented this

File: src/Components/Search/search.png
Comment (alivaditis) [reply src/Components/Search/search.png:1]: cool, i'll move it in there and change the import when merges happen :)

File: src/Components/App/App.js (hunk 2 above)
Comment (alivaditis) [reply src/Components/App/App.js]: nice catch, upon thinking about this more I got rid of the query state and used useParams instead, now the url reflects the query and the page, this fixed the refresh bug

File: src/Components/Results/Results.js (hunk 3 above)
Comment (alivaditis) [reply src/Components/Results/Results.js:22]: Yeah, this was a sort of makeshift solution until the other components are in place/ we decide where that error state should go.

File: src/Components/Search/Search.css (hunk 7 above)
Comment (alivaditis) [reply src/Components/Search/Search.css:8]: Yes, at least for mobile first design.  We could discuss what it would look like on larger screens.

Comment (jmartin777) [reply src/Components/Search/Search.css:8]: Definitely would be great discussion to address as we start implementing Responsive dimensions to each state and keeping that uniform throughout the app!

File: src/Components/Search/search.png
Comment (jmartin777) [reply src/Components/Search/search.png:1]: i made an Assets folder (located in src directory) in my push as well .

File: src/Components/Results/Results.js (hunk 5 above)
Comment (jmartin777) [reply src/Components/Results/Results.js]: nice perspective on the wording 👍

**Review by jmartin777 — COMMENTED** [review]
Great job Alec on testing and mapping out these fetch calls for searchand testing them !

File: src/Components/Results/Results.css (hunk 4 above)
Comment (alivaditis) [reply src/Components/Results/Results.css:20]: fixed!

File: src/api.js (hunk 1 above)
Comment (alivaditis) [reply src/api.js]: added new handleError function to the api file!

**Review by jalbe0076 — APPROVED** [review]
Looking good!

File: src/Components/Card/Card.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import React from 'react'
+import { Link } from 'react-router-dom'
+import './Card.css'
+
+function Card({result}) {
+  return ((<Link to= {`/albums/${result.masterId}`} className='results--card' id={result.masterId}>
+    <img className='results--image' src={result.thumb}/>
+    <p className='results--title'>{result.title.length > 20 ? result.title.slice(0, 20) + '...' : result.title}</p>
+    <p className='results--artist'>{result.artist.length > 20 ? result.artist.slice(0, 20) + '...' : result.artist}</p>
+  </Link>))
```
Comment (mbenfowler) [inline src/Components/Card/Card.js:10]: Just some minor formatting

```suggestion
  return (
    <Link to={`/albums/${result.masterId}`} className='results--card' id={result.masterId}>
      <img className='results--image' src={result.thumb}/>
      <p className='results--title'>{result.title.length > 20 ? result.title.slice(0, 20) + '...' : result.title}</p>
      <p className='results--artist'>{result.artist.length > 20 ? result.artist.slice(0, 20) + '...' : result.artist}</p>
    </Link>
  )
```

File: src/Components/Results/Results.css [hunk 10]
```diff
@@ -0,0 +1,100 @@
+.results {
+  background-color: rgb(20, 18, 17);
```
Comment (mbenfowler) [inline src/Components/Results/Results.css:2]: any reason you chose to use the `rgb` rather than the hex?

File: src/Components/Results/Results.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import React, {useState, useEffect} from 'react'
+import { useParams, Link } from 'react-router-dom'
+import Card from '../Card/Card'
+import './Results.css'
+import { searchAlbums } from '../../api'
+
+function Results() {
+  const {query, page} =useParams()
```
Comment (mbenfowler) [inline src/Components/Results/Results.js:8]: space after equals here

File: src/components/App/App.js [hunk 12]
```diff
@@ -5,17 +5,23 @@ import Discover from '../Discover/Discover';
...
       <Nav />
+      <Search/>
       <main className="App">
         <Routes>
           <Route path="/" />
           <Route path="/journal" element={<Journal />} />
           <Route path="/collections" element={<Collections />} />
           <Route path="/discover" element={<Discover />} />
+          <Route path='/search/:query/:page' element={<Results/>}/>
+          <Route path='/albums/:id' element={<Album/>}/>
```
Comment (mbenfowler) [inline src/components/App/App.js:24]: Love all the work you did with pagination to handle for this route, but I'd like some discussion on whether we need/want individual routes for each page.  The only benefit I see is it allowing use of the back/forward browser buttons - wouldn't it be easier if we just didn't route this stuff and expect the user to only use the provided `prev`/`next` buttons?  It would def make for cleaner code in our `Results` component.

@jmartin777 @jalbe0076 would love ya'lls thoughts too on this

**Review by mbenfowler — CHANGES_REQUESTED** [review]
A couple of things to look at!

File: src/Components/Card/Card.css [hunk 13]
```diff
@@ -0,0 +1,28 @@
+.results--card {
```
[the remark this answers is not in the document]
Comment (alivaditis) [reply src/Components/Card/Card.css:1]: I saw this on the scrimba course.  I didn't want to have a situation where it was '.card' and maybe somebody else has a separate card component resulting in conflicting styling, just a way of making sure it's unique.

File: src/components/App/App.js (hunk 12 above)
Comment (jalbe0076) [reply src/components/App/App.js:24]: I like the idea of being able to use the forward and back buttons. I personally use it all the time and opt for it rather than the back button on a page.

File: src/Components/Card/Card.js (hunk 9 above)
Comment (alivaditis) [reply src/Components/Card/Card.js:10]: fixed

File: src/Components/Results/Results.css (hunk 10 above)
Comment (alivaditis) [reply src/Components/Results/Results.css:2]: not really, i was using the digital color meter that comes with macs to get the color before it was updated in the wireframe, and that uses rgb by default

File: src/components/App/App.js (hunk 12 above)
Comment (alivaditis) [reply src/components/App/App.js:24]: I tried that first but I was running into an error where the app broke when refreshing.
Basically you need the page in the url because if you're on a specific page and refresh, it breaks.  It seems like this is the way it's done in the real world, if you look at the url when going through pages of results on websites it's usually in the url somehow.  By using useParams on load of the url to get page and query, the api call sets results and pagination information and the state is all in the results file.  Before I was passing around setter functions and states, and now it's all in results, so it makes it cleaner in that way.

Comment (mbenfowler) [reply src/components/App/App.js:24]: Okay great! Thanks for the discussion yall and for putting all the thought and experimentation into this Alec - pagination is always tricky!

**Review by mbenfowler — APPROVED** [review]
